return function()
	local ReplicatedStorage = game:GetService("ReplicatedStorage")
    local Connection = require(ReplicatedStorage.Modules.Connection)
	local NetworkValue = Connection.NetworkValue

    describe("NetworkValueClient", function()
		it("Should create a new network value", function()
			local value = NetworkValue.new("TestValue1", ReplicatedStorage)
			expect(value).to.be.ok()
			expect(value.className).to.equal("NetworkValue")
		end)
		it("Should be destroyable", function()
			local value = NetworkValue.new("TestValue2", ReplicatedStorage)
			value:Destroy()
			expect(value._NetworkEvent).to.equal(nil)
		end)
		it("Should override local value when server updates it", function()
			local value = NetworkValue.new("TestValue3", ReplicatedStorage, 123)
			expect(value:Get()).to.never.equal(123)
			expect(value:WaitAsync()).to.equal(500)
		end)
		it("Should fire connection when value is changed server-side", function()
			local value = NetworkValue.new("TestValue4", ReplicatedStorage)
			local changed = false
			local connection = value:Connect(function(newValue)
				changed = true
				expect(newValue).to.equal(789)
			end)
			task.delay(1, function()
				expect(changed).to.equal(true)
				connection:Disconnect()
			end)
		end)
	end)
end